pkg/util/json/json.go

```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
 	return jsonObject(append([]jsonKeyValuePair(nil), b.pairs...)), nil
 }
 
+// PartialObject is a JSON object builder that builds an object using a set of
+// fixed key-values (supplied at construction) and a set of keys which are
+// settable when building the final JSON object.
+type PartialObject struct {
+	basePairs      []jsonKeyValuePair
+	keyOrd         map[string]int
+	allowedNewKeys []string
+}
+
+// NewPartialObject creates a PartialObject with the specified base key-value
+// pairs, and allows for setting the values of newKeys when building the final
+// JSON object.
+func NewPartialObject(base map[string]JSON, newKeys []string) (*PartialObject, error) {
+	// Check that newKeys are unique.
+	newKeySet := make(map[string]struct{}, len(newKeys))
+	for _, k := range newKeys {
+		if _, ok := newKeySet[k]; ok {
+			return nil, errors.AssertionFailedf("expected unique keys, found %v", newKeys)
+		}
+		newKeySet[k] = struct{}{}
+	}
+
+	sort.Strings(newKeys)
+
+	keys := slices.Collect(maps.Keys(base))
+	keys = append(keys, newKeys...)
+	sort.Strings(keys)
+
+	pairs := make([]jsonKeyValuePair, 0, len(base))
+	keyOrd := make(map[string]int, len(keys))
+
+	for i, k := range keys {
+		keyOrd[k] = i
+		if _, ok := base[k]; !ok {
+			continue
+		}
+		pairs = append(pairs, jsonKeyValuePair{k: jsonString(k), v: base[k]})
+	}
+
+	return &PartialObject{basePairs: pairs, keyOrd: keyOrd, allowedNewKeys: newKeys}, nil
+}
+
+// NewObject creates a new JSON object with the specified new key-values. The
+// order of newData MUST match the order of newKeys supplied to this
+// PartialObject's constructor.
+func (po *PartialObject) NewObject(newData []JSON) (JSON, error) {
```

Comment:
I wonder if it would be less error prone to have this function accept a `map[string]JSON`

Reply:
yeah i was going back and forth on this myself. i'll change it, thx for the input